### ISSUE TITLE
Protect mutations on @connection based on ownerField in source object

### ISSUE THREAD

**Issue by jroebu14** [aws-amplify/amplify-cli#7239]
**Which Category is your question related to?**
GraphQL API

**Amplify CLI Version** 
4.50.0

**What AWS Services are you utilizing?**
AppSync

**Provide additional details e.g. code snippets**
Assuming I have this schema:

```graphql
type Organiser
  @model
  @auth(rules: [{ allow: owner }, { allow: owner, ownerField: "admins" }]) {
  id: ID!
  admins: [ID]
  name: String!
  events: [Event] @connection(name: "OrganiserEvents")
}

type Event
  @model
  @auth(rules: [{ allow: public, operations: [read] }]) {
  id: ID!
  name: String!
  organiser: Organiser @connection(name: "OrganiserEvents")
}
```

How might I prevent users from creating/updating/deleting an `Event` unless they belong to the `owner` or `admins` field in `Organiser`?

The closest example I've found is around field-level authorization, for example in the section `Protect mutations such that certain fields can have different access rules than the parent model` on this page https://docs.amplify.aws/cli/graphql-transformer/auth#field-level-authorization but this of course would not work for a field with a `@connection` resolver.

Many thanks!

**Comment by attilah** [comment]
@jroebu14 I don't know if that was clear from the docs or now, but the key is to disable the `queries` on the child model `Event` since that way you can only access `Events` through the parent `Organiser` so ownership based protection will be enforced in the generated resolvers.

I see that you require `public` `read` access to `Events` directly, is that something you require or going through `Organiser` would also work to get those records?

**Comment by jroebu14** [comment]
@attilah thank you for getting back to me.

> I don't know if that was clear from the docs or now, but the key is to disable the queries on the child model Event

Oh ok. Tbh I read that as an optional thing that you could do rather than something that is required for ownership based protection to be enforced in the generated resolvers. That's very cool. Thanks!

> I see that you require `public` `read` access to `Events` directly, is that something you require or going through `Organiser` would also work to get those records?

Going through the `Organiser` wouldn't be ideal in my case. Is there a way to also enable `public` `read` access to `Events` directly?

**Comment by jroebu14** [comment]
@attilah while I waited on an answer to my question I got things working by disabling `mutations` on the child model `Event` and adding a `createEvent` mutation type with a custom lambda resolver that checks the `admins` field on `Organiser`.

e.g.

```graphql
type Mutation {
  createEvent(
    id: ID!
    name: String!
    eventOrganiserID: ID!
  ): Event @function(name: "events-${env}")
}
```

**Comment by attilah** [comment]
Using `@function` is always a good escape hatch / solution since you get the full control which has its pros/cons of course, if that is the preferred solution it will fully work for you.

Regarding querying `Events` directly, you can add a named `@key` directive, which will generated a query for you and you can use the referenced key arguments to do filtering/sorting.

**Comment by jroebu14** [comment]
> Regarding querying `Events` directly, you can add a named `@key` directive, which will generated a query for you and you can use the referenced key arguments to do filtering/sorting.

@attilah makes sense. Thanks very much that's really useful. I'll give this a try tomorrow 🙂 

I think I'd end up with something like this then:

```graphql
type Organiser
  @model
  @auth(rules: [{ allow: owner }, { allow: owner, ownerField: "admins" }]) {
  id: ID!
  admins: [ID]
  name: String!
  events: [Event]
    @connection(name: "OrganiserEvents")
    @auth(rules: [{ allow: owner, ownerField: "admins" }])
}

type Event
  @model(queries: null)
  @key(
    name: "byStartDate"
    fields: ["queryName", "startDate"]
    queryField: "eventsByStartDate"
  )
  @auth(rules: [{ allow: public, operations: [read] }]) {
  id: ID!
  queryName: String!
  name: String!
  organiser: Organiser @connection(name: "OrganiserEvents")
}
```

**Comment by jroebu14** [comment]
One last question on this issue I have is around having the same ownership based protection from `Organiser` enforced on child models of `Event`.

For example, see the child model of `Event` called `Things`:

```graphql
type Organiser
  @model
  @auth(rules: [{ allow: owner }, { allow: owner, ownerField: "admins" }]) {
  id: ID!
  admins: [ID]
  name: String!
  events: [Event]
    @connection(name: "OrganiserEvents")
    @auth(rules: [{ allow: owner, ownerField: "admins" }])
}

type Event
  @model(queries: null)
  @key(
    name: "byStartDate"
    fields: ["queryName", "startDate"]
    queryField: "eventsByStartDate"
  )
  @auth(rules: [{ allow: public, operations: [read] }]) {
  id: ID!
  queryName: String!
  name: String!
  organiser: Organiser @connection(name: "OrganiserEvents")
  things: [Thing] @connection(name: "EventThings") # I would like this protected by the ownerField "admins" in Organiser
}
```

How would this be possible? Would I also have to add the child model of `Event` to `Organiser` with the `@auth` directive?

**Comment by jroebu14** [comment]
> > Regarding querying `Events` directly, you can add a named `@key` directive, which will generated a query for you and you can use the referenced key arguments to do filtering/sorting.
> 
> @attilah makes sense. Thanks very much that's really useful. I'll give this a try tomorrow 🙂
> 
> I think I'd end up with something like this then:
> 
> ```graphql
> type Organiser
>   @model
>   @auth(rules: [{ allow: owner }, { allow: owner, ownerField: "admins" }]) {
>   id: ID!
>   admins: [ID]
>   name: String!
>   events: [Event]
>     @connection(name: "OrganiserEvents")
>     @auth(rules: [{ allow: owner, ownerField: "admins" }])
> }
> 
> type Event
>   @model(queries: null)
>   @key(
>     name: "byStartDate"
>     fields: ["queryName", "startDate"]
>     queryField: "eventsByStartDate"
>   )
>   @auth(rules: [{ allow: public, operations: [read] }]) {
>   id: ID!
>   queryName: String!
>   name: String!
>   startDate: String!
>   organiser: Organiser @connection(name: "OrganiserEvents")
> }
> ```

I tried this out but unfortunately I get an error `Not Authorized to access createEvent on type Mutation`.

**Comment by attilah** [comment]
I don't see why you need 2 owner rules for Organiser and also no need for the `@auth` on the field as fields are inheriting that from the type level even for connections IIRC. Breaking down the root issue: `Event` does not have an owner field, so that is like missing data to enforce ownership. Another approach would be to manage the two entities separately.

First I'd make ownership working correctly and then enhance the schema with modifications for `@key` and read only `public` access.

```graphql
type Organiser
  @model
  @auth(rules: [{ allow: owner, ownerField: "admins" }]) {
  id: ID!
  admins: [ID]
  name: String!
  events: [Event]
    @connection(name: "OrganiserEvents")
}

type Event
  @model(queries: null)
  id: ID!
  queryName: String!
  name: String!
  startDate: String!
  organiser: Organiser @connection(name: "OrganiserEvents")
}
```

Check if this would work for your use case.

**Comment by jroebu14** [comment]
Thanks for your example schema @attilah

Unfortunately I don't think CreateEvent mutations are protected though. This is how I tested after making the schema changes and pushing:

- I go to the AWS Console in AppSync queries UI
- I log in with Cognito User Pools (account A)
- I create an Organiser and then create an Event with the generated `eventOrganiserId`
- I query with listOrganisers and can see 1 event in the `events` field
- I logout of account A and login with a different account (account B)
- I am able to create an Event using the same `eventOrganiserId`
- I query with listOrganisers and the `events` field on Organiser is empty
- I logout of account B and login with account A
- I query with listOrganisers and can see 2 events in the `events` field

I would expect not to be able to create an Event with account B using the `eventOrganiserId` that was generated by creating an Organiser with account A and to then see this event when logged in with account A

I may be completely misunderstanding something though.

**Comment by jroebu14** [comment]
@attilah `eventOrganiserId ` is a database ID.

I can confirm that the user's username is being stored in the `admins` array.

**Comment by attilah** [comment]
this part:
> I create an Organiser and then create an Event with the generated eventOrganiserId
is not enforced by the resolver so the test case per se is not right as this is a database id and it has no 1:1 relation to the authenticated user the generated resolver is not enforcing any of that during createEvent.
